libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/SchemaValidationScope.kt

```diff
@@ -433,9 +434,30 @@ internal fun ValidationScope.validateRootOperationTypes(schemaDefinition: GQLSch
 }
 
 private fun ValidationScope.validateInterfaces() {
-  typeDefinitions.values.filterIsInstance<GQLInterfaceTypeDefinition>().forEach {
-    if (it.fields.isEmpty()) {
-      registerIssue("Interfaces must specify one or more fields", it.sourceLocation)
+  typeDefinitions.values.filterIsInstance<GQLInterfaceTypeDefinition>().forEach { i ->
+    if (i.fields.isEmpty()) {
+      registerIssue("Interfaces must specify one or more fields", i.sourceLocation)
+    }
+
+    i.implementsInterfaces.forEach { implementsInterface ->
+      val iface = typeDefinitions[implementsInterface] as? GQLInterfaceTypeDefinition
+      if (iface == null) {
+        registerIssue("Interface '${i.name}' cannot implement non-interface '$implementsInterface'", i.sourceLocation)
+      }
+    }
+
+    i.directives.forEach { directive ->
+      validateDirective(directive, i) {
+        issues.add(it.constContextError())
+      }
+    }
+
+    i.fields.forEach { gqlFieldDefinition ->
+      gqlFieldDefinition.directives.forEach { gqlDirective ->
+        validateDirective(gqlDirective, gqlFieldDefinition) {
+          issues.add(it.constContextError())
```

Comment:
FWIW, this is where having a visitor pattern for validation would be useful. Maybe not right now but at some point.